core/src/main/scala/spinal/core/sim/SpinalSpec.scala

```diff
@@ -0,0 +1,276 @@
+  val defaultCaching: Boolean = true
+
+  /** Define a group of tests that can be run on several benches */
+  case class TestGroup[SoftDut](tests: Bench[SoftDut] => Unit) extends Nameable {
+
+    /** Run the tests on (a) named (in a single `val`) benche(s) */
+    def runWith(benches: Bench[SoftDut]*): Unit = {
+      getName - {
+        for (bench <- benches) {
+          bench.getName - {
+            tests(bench)
+          }
+        }
+      }
+    }
+
+    /** Run the tests on a sequence of benches */
+    def runWithAll(namedBenches: Seq[(Bench[SoftDut], String)]): Unit = {
+      getName - {
+        for ((bench, name) <- namedBenches) {
+          name - {
+            tests(bench)
+          }
+        }
+      }
+    }
+  }
+
+  /** Define a group of tests that can be run with several component configurations
+    *
+    * {{{
+    * val myTests =
+    *  TestGroupCfg { (cfg: MyComponentConfig) =>
+    *    bench(new MyComponent(cfg))
+    *  } { (cfg, it) =>
+    *    it should "do nothing" in { dut => }
+    *  }
+    * }}}
+    *
+    * @param benchBuilder the function to create a bench from the component configuration
+    * @param tests tests to run, which can be conditional over the component configuration
+    */
+  case class TestGroupCfg[Cfg, SoftDut](benchBuilder: Cfg => Bench[SoftDut])(
+      tests: (Cfg, Bench[SoftDut]) => Unit
+  ) extends Nameable {
+
+    /** Override this function to define how configurations should appear in test results */
+    def cfgToString(cfg: Cfg): String = cfg.toString()
+
+    /** Run the tests with (a) component configuration(s) */
+    def runWith(cfgs: Cfg*): Unit = runWithAll(cfgs)
+
+    /** Run the tests with a sequnece of component configurations */
+    def runWithAll(cfgs: Seq[Cfg]): Unit = {
+      getName - {
+        for (cfg <- cfgs) {
+          cfgToString(cfg) - {
+            tests(cfg, benchBuilder(cfg))
+          }
+        }
+      }
+    }
+  }
+
+  /** Create a test-bench
+    *
+    * @param dut The device to test
+    * @param preSimHook An action to do on `dut` before each test (defaults to nothing)
+    * @param finalConfig To override `SpinalSpec.config`
+    * @param finalCaching To override `SpinalSpec.caching`
+    */
+  def bench[Dut <: Component](
+      dut: => Dut,
+      preSimHook: Dut => Unit = { (_: Dut) => },
+      finalConfig: SpinalSimConfig = defaultConfig,
+      finalCaching: Boolean = defaultCaching
+  ): Bench[Dut] = {
+    benchTransform(dut, { (dut: Dut) => preSimHook(dut); dut }, finalConfig, finalCaching)
+  }
+
+  /** Create a test-bench with an abstraction of the `dut`
+    *
+    * @param dut The device to test
+    * @param preSimTransform Function taking the `dut` and returning your abstracted dut
+    * @param finalConfig To override `SpinalSpec.config`
+    * @param finalCaching To override `SpinalSpec.caching`
+    */
+  def benchTransform[HardDut <: Component, SoftDut](
```

Comment:
Sorry for the long reply...

To use-case for this function to me initially seemed it was to allow writing kind of generic tests for two implementations that "do the same thing" but have different implementations, i.e. a parallel adder and a bitserial adder (just to explain what I mean).

Something like:
```
case class ParallelAdder() ... { val io = { val a = UInt(...), ... }}
case class SerialAdder() ... { val io = { val a = Bool(), ... }}
abstract class Adder { def doAdd(a: Int, b: Int): Int = ??? }
case class ParallelAdderWrapper(dut: ParallelAdder) extends Adder { 
  override def add(a: Int, b: Int): Int = { dut.io.a #= a; ...; sleep(1); dut.io.result.toInt }
}
case class SerialAdderWrapper(dut: SerialAdder) extends Adder {
  override def add(a: Int, b: Int): Int = {
    for(i <- a.range) {
      // set one bit here
      dut.clockDomain.waitSampling()
      // get one bit here
    }
    // return final result
}
def makeAdder(dut: SerialAdder): Adder = SerialAdderWrapper(dut)
def makeAdder(dut: ParallelAdder): Adder = ParallelAdderWrapper(dut)

...

benchTransform(?dut?, makeAdder) { it => 
  it should "add correctly" {
    assert(it.add(24, 18), 42)
  }
}
```
But that doesn't work since one still needs to pass the DUT... (kind of a DUT-independent TestGroup)
So I'm wondering what use case you had in mind, because currently it looks to me like a complicated way of adding a helper function that we could otherwise get by just making a function that gets the DUT as a parameter, or if we really wanted the syntax use an implicit class? You call it "software abstraction" in a few places, but I'm not sure what you exactly had in mind there.

PS: not saying it should do what I initially though... ;-)

Reply:
The initial use-case of `benchTransform` is for things like (draft code, didn't check buses names XD):

```scala
case class Ahb2ApbBridge() extends Component {
  val io = new Bundle {
    val ahb = slave  port AhbLite(...)
    val apb = master port Apb3(...)
  }
  ...
}

case class SoftAhb2ApbBridge(dut: Ahb2ApbBridge) {
  val ahb = AhbDriver(dut.io.ahb)
  val apb = ApbSlaveSubstitute(dut.io.apb)
}

val bridge = benchTransform(Ahb2ApbBridge(), SoftAhb2ApbBridge)

bridge should "read" { dut =>
  val read = dut.ahb.Read(0xaaaa)
  dut.apb.setOnReads(addr => addr + 1)
  read.run()
  assert(read.value == 0xaaab)
  assert(read.waitCycles = ...)
}
```

Then in a test you have helpers to emulate a slave (with `ApbSlaveSubstitute`) and a master (with `AhbDriver`) so that you can write tests with a transaction-level approach (#879), for instance perform a read at given address with given attributes, await the answer, assert the read value and the number of inserted wait states, etc.

Tbh I'm wondering if the transformation (along with `preSimHook` stuff) should be performed on test groups instead of benches because:

* the hardware compiled is actually the same
* it puts the code run before tests closer to the code of the tests

The drawback would be that it would not be possible to use this feature outside of a test group.

@andreasWallner What do you think about that?

---

### Below I did not notice the two components didn't have the same interface, you can ignore it

Hmm when I was designing the API, I wanted to address a use-case which looks like yours. Below I try to transpose your use case into what I am addressing, tell me if it matches for you:

```scala
// only io interface, no logic here
abstract class Adder extends Component { val io = new Bundle { ... } }
// logic, both using io defined earlier
case class SerialAdder() extends Adder { ... }
case class ParallelAdder() extends Adder { ... }
// software abstraction which is an interface which does not depend on the component implementation
// the main use-case here is to define protocols to perform transaction-level tests
case class SoftAdder(dut: Adder) {
  def add(a: Int, b: Int): Int = ...
  ...
}

val serialAdder = benchTransform(SerialAdder(), SoftAdder)
val parallelAdder = benchTransform(ParallelAdder(), SoftAdder)

val myTestGroup = TestGroup[SoftAdder] { it =>
  it should "perform additions" inSim { dut =>
    assert(dut.add(24, 18), 42)
  }
}
```

Reply:
Ah, ok - that makes more sense than what I'd though about. From the descriptions/names I always though about providing a software abstraction for the DUT (e.g. you call it `SoftAhb2ApbBridge` which kind of implies to me that it is a SW version of the bridge; similar with `SoftDut`, whereas the `Soft...` part actually contains part of the testbench).

In that form it makes sense to me, containing drivers, scoreboards, etc. But: that also means that all of those are then accessed via `dut. ...`, but those drivers and scoreboards are not part of the DUT, they are part of the testbench. In your example it reads nice, but in other cases this might range from weird (`dut.rxScoreboard.pushRef(...)`) to confusing (was it `dut.ahb...` or `dut.dut.ahb...`?)

Regarding location: I see why you'd want to move the stuff to a `TestGroup`, but that would also mean that w/o one you can't use it that way (Not sure that'd be a big problem). Come to think of it: For the use-case you described, you could just use the case-class w/o the `benchTransform` - just instantiate it in either a `test`, after the `bench` statement or in a `TestGroup`. That should work just the same, just with the dut and testbench being two separate objects? (and leaves all options open)

Small comment on naming: `Bench` might be easily misunderstood, I checked with some of my digital designer colleagues and they would expect a similar meaning to what Wikipedia mentions:
> The term "test bench" is used in digital design with a [hardware description language](https://en.wikipedia.org/wiki/Hardware_description_language) to describe the test code, which instantiates the DUT and runs the test.
Where only the first part is true here, not the second. (actually less true if you do the move ;-) )
But I see why you used it, and I don't have an alternative I'm really happy with...